sherpa/conftest.py

```diff
@@ -589,66 +617,93 @@ def old_numpy_printing():
 
 
-PLOT_BACKENDS = [dummy_backend]
-if pylab_backend is not None:
+@pytest.fixture(params=PLOT_BACKENDS.keys())
+def all_plot_backends(request):
+    """Override the plot backend for this test
 
-    # override the matplotlib backend from TkAgg to Agg to try
-    # and avoid errors from multiprocessing thanks to the error
-    # "Tcl_AsyncDelete: async handler deleted by the wrong thread"
-    # See issue #1590.
-    #
-    import matplotlib
-    matplotlib.use("agg")
+    Runs the test with the given plot backend and then restores the
+    original value.
 
-    PLOT_BACKENDS.append(pylab_backend)
+    Note that this does not reload the ui layers, which will
+    retain a reference to the original backend throughout.
+    """
+    with TemporaryPlottingBackend(request.param):
+        yield
+        if request.param in ('pylab', 'PylabErrorArea'):
+            from matplotlib import pyplot as plt
+            plt.close(fig="all")
 
 
-@pytest.fixture(params=PLOT_BACKENDS)
-def override_plot_backend(request):
-    """Override the plot backend for this test
+@pytest.fixture(params=PLOT_BACKENDS.keys())
+def all_plot_backends_astro_ui(request):
+    """Override the plot backend for the astro ui session.
 
     Runs the test with the given plot backend and then restores the
     original value.
-
-    Note that this does not reload the ui layers, which will have
-    retain reference to the original backend throughout.
-
     """
 
-    old = sherpa.plot.backend
+    # Unfortunately can not easily use TemporaryPlottingBackend here
+    #
+    from sherpa.plot import backend
+    old = backend
 
-    changed = old.name != request.param.name
-    if changed:
-        sherpa.plot.backend = request.param
-        sherpa.astro.plot.backend = request.param
+    from sherpa.astro.ui import set_plot_backend
+    set_plot_backend(request.param)
 
     yield
-    if changed:
-        sherpa.plot.backend = old
-        sherpa.astro.plot.backend = old
+
+    # Backend-specific clean up.
+    if request.param == 'pylab':
+        from matplotlib import pyplot as plt
+        plt.close(fig="all")
+
+    # Restore the original backend.
+    #
+    set_plot_backend(old)
+
+
+@pytest.fixture(params=FUNCTIONAL_PLOT_BACKENDS)
+def plot_backends(request):
+    """Override the plot backend for this test (only for functional backends)
+
+    This fixtures runs tests for all backends that actually produce plots, i.e.
+    not for any type of dummy.
+
+    Runs the test with the given plot backend and then restores the
+    original value.
+
+    Note that this does not reload the ui layers, which will
+    retain a reference to the original backend throughout.
+    """
+
+    with TemporaryPlottingBackend(request.param):
+        yield
+        if request.param == 'pylab':
+            from matplotlib import pyplot as plt
+            plt.close(fig="all")
 
 
 @pytest.fixture(autouse=True, scope="session")
 def cleanup_pylab_backend():
```

Comment:
Thinking aloud: this fixture should now only be needed/used with tests that have the `requires_pylab` decorator. If so, is it worth trying to combine them? Probably not, since this is session scoped rather than fixture scoped. It would be different if we pulled out all the pylab tests into their own test files, but this is not relevant for this PR.